File: tools/native/rust/stunir-native/src/toolchain.rs
use serde::Deserialize;
use std::collections::BTreeMap;
use std::fs;
use std::path::Path;
use std::process::exit;
use crate::canonical::sha256_hex;

#[derive(Deserialize, Debug)]
struct ToolDef {
    path: String,
    sha256: String,
    version: String,
}

type ToolchainLock = BTreeMap<String, ToolDef>;

pub fn verify_toolchain(lock_path: &str) {
    if !Path::new(lock_path).exists() {
        eprintln!("Error: Lockfile not found at {}", lock_path);
        exit(1);
    }

    let content = fs::read_to_string(lock_path).expect("Failed to read lockfile");
    let tools: ToolchainLock = match serde_json::from_str(&content) {
        Ok(t) => t,
        Err(_) => {
            eprintln!("Error: Failed to parse toolchain lockfile.");
            exit(1);
        }
    };

    println!("Verifying {} tools from {}", tools.len(), lock_path);

    for (name, def) in tools {
        verify_tool(&name, &def);
    }
    println!("Toolchain verification passed.");
}

fn verify_tool(name: &str, def: &ToolDef) {
    if !Path::new(&def.path).exists() {
        eprintln!("Error: Tool '{}' not found at {}", name, def.path);
        exit(1);
    }

    let content = fs::read(&def.path).expect("Failed to read tool binary");
    let actual_hash = sha256_hex(&content);

    if actual_hash != def.sha256 {
        eprintln!("Error: Hash mismatch for tool '{}'", name);
        eprintln!("  Path: {}", def.path);
        eprintln!("  Expected: {}", def.sha256);
        eprintln!("  Actual:   {}", actual_hash);
        exit(1);
    } else {
        println!("  [OK] {}", name);
    }
}

File: docs/archive/native_legacy/native_tools/haskell/stunir-native/src/STUNIR/Toolchain.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}
module Stunir.Toolchain where

import GHC.Generics
import Data.Aeson
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.IO as TIO
import qualified Data.ByteString.Lazy as BL
import qualified Data.Map.Strict as Map
import System.Directory (doesFileExist)
import Control.Monad (forM_, when)
import System.Exit (exitFailure)
import Crypto.Hash.SHA256 (hash)
import qualified Data.ByteString.Base16 as Hex
import qualified Data.Text.Encoding as TE

-- | Tool definition in lockfile
data ToolDef = ToolDef {
    td_path :: Text,
    td_sha256 :: Text,
    td_version :: Text
} deriving (Show, Eq, Generic)

instance FromJSON ToolDef where
    parseJSON = withObject "ToolDef" $ \v -> ToolDef
        <$> v .: "path"
        <*> v .: "sha256"
        <*> v .: "version"

-- | The Lockfile is a Map of ToolName -> ToolDef
type ToolchainLock = Map.Map Text ToolDef

-- | Helper to compute SHA256 of a file
hashFile :: FilePath -> IO Text
hashFile path = do
    content <- BL.readFile path
    let h = hash (BL.toStrict content)
    return $ TE.decodeUtf8 (Hex.encode h)

-- | Verify the toolchain
verifyToolchain :: FilePath -> IO ()
verifyToolchain lockPath = do
    exists <- doesFileExist lockPath
    if not exists then do
        putStrLn $ "Error: Lockfile not found at " ++ lockPath
        exitFailure
    else do
        content <- BL.readFile lockPath
        case decode content :: Maybe ToolchainLock of
            Nothing -> do
                putStrLn "Error: Failed to parse toolchain lockfile."
                exitFailure
            Just tools -> do
                putStrLn $ "Verifying " ++ show (Map.size tools) ++ " tools from " ++ lockPath

                -- Iterate over tools
                mapM_ verifyTool (Map.toList tools)

                putStrLn "Toolchain verification passed."

verifyTool :: (Text, ToolDef) -> IO ()
verifyTool (name, def) = do
    let path = T.unpack (td_path def)
    let expectedHash = td_sha256 def

    -- 1. Check Existence
    exists <- doesFileExist path
    if not exists then do
        putStrLn $ "Error: Tool '" ++ T.unpack name ++ "' not found at " ++ path
        exitFailure
    else do
        -- 2. Check Hash
        -- Note: In a real large-scale build, we might skip hashing for performance unless strict mode is on.
        -- For STUNIR, correctness is paramount, so we hash.
        actualHash <- hashFile path

        if actualHash /= expectedHash then do
            putStrLn $ "Error: Hash mismatch for tool '" ++ T.unpack name ++ "'"
            putStrLn $ "  Path: " ++ path
            putStrLn $ "  Expected: " ++ T.unpack expectedHash
            putStrLn $ "  Actual:   " ++ T.unpack actualHash
            exitFailure
        else do
            putStrLn $ "  [OK] " ++ T.unpack name
